### PR TITLE
[AMDGPU] Fix end() iterator dereference in SILowerSGPRSpills

### PR DESCRIPTION
**Pull request by bcahoon**

### CODE REVIEW

**Comment by bcahoon** [comment]
I added a test case that shows the error. I'll try to reduce it further, though I'm not sure how useful it is.

File: llvm/test/CodeGen/AMDGPU/sgpr-spills-empty-block.ll [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+67:                                               ; preds = %48
+  br label %.sink.split
+
+68:                                               ; preds = %48
+  br label %.sink.split
+
+69:                                               ; preds = %48
+  br label %.sink.split
+
+70:                                               ; preds = %48
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/sgpr-spills-empty-block.ll:463]: Use named values

**Review by arsenm — COMMENTED** [review]
This needs to be a MIR test to be reliable

File: llvm/test/CodeGen/AMDGPU/sgpr-spills-empty-block.ll (hunk 1 above)
Comment (bcahoon) [reply llvm/test/CodeGen/AMDGPU/sgpr-spills-empty-block.ll:463]: Still trying to reduce the mir version - with this ll input, the mir is about 1700 lines long.

File: llvm/lib/Target/AMDGPU/SILowerSGPRSpills.cpp [hunk 2]
```diff
@@ -278,7 +278,10 @@ void SILowerSGPRSpills::extendWWMVirtRegLiveness(MachineFunction &MF,
   for (auto Reg : MFI->getSGPRSpillVGPRs()) {
     for (MachineBasicBlock *SaveBlock : SaveBlocks) {
       MachineBasicBlock::iterator InsertBefore = SaveBlock->begin();
-      auto MIB = BuildMI(*SaveBlock, *InsertBefore, InsertBefore->getDebugLoc(),
+      DebugLoc DL = InsertBefore != SaveBlock->end()
+                        ? InsertBefore->getDebugLoc()
+                        : DebugLoc();
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SILowerSGPRSpills.cpp:283]: This looks like a job for `MachineBasicBlock::findDebugLoc` (not that I've ever used it myself):
```suggestion
      DebugLoc DL = SaveBlock->findDebugLoc(InsertBefore);
```

Comment (bcahoon) [reply llvm/lib/Target/AMDGPU/SILowerSGPRSpills.cpp]: Thanks - findDebugLoc is much better.

File: llvm/test/CodeGen/AMDGPU/sgpr-spills-empty-block.mir [hunk 3]
```diff
@@ -0,0 +1,1445 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
+# RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx942 -verify-machineinstrs -run-pass=si-lower-sgpr-spills -o - %s | FileCheck %s
+
+# Test that the si-lower-sgpr-spills pass does not defeference a
+# machine instruction iterator that is equal to end().
+
+---
+name:            empty_spill_block
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/sgpr-spills-empty-block.mir:8]: This test function is huge. Can you try to use llvm-reduce to cut this down? I can't imagine needing more than a few blocks

Comment (bcahoon) [reply llvm/test/CodeGen/AMDGPU/sgpr-spills-empty-block.mir:8]: Yeah, I've been trying to create a smaller version. This one was generated with llvm-reduce. I don't really like the test size either.

**Comment by cdevadas** [comment]
Yes, `findDebugLoc` is the right one.

A simple case like the following would be enough.
Leave the first BB empty and a spill at any successor block would suffice. 
The test name 'empty_spill_block' is misleading. May be, empty_prolog_block?
---
name:            empty_prolog_block
stack:
  - { id: 0, name: '', type: spill-slot, offset: 0, size: 32, alignment: 4,
      stack-id: sgpr-spill, callee-saved-register: '', callee-saved-restored: true,
      debug-info-variable: '', debug-info-expression: '', debug-info-location: '' }
machineFunctionInfo:
  hasSpilledSGPRs: true
body:             |
  bb.0:
    liveins: $sgpr0
  bb.1:
    renamable $sgpr0 = SI_SPILL_S32_RESTORE %stack.0, implicit $exec, implicit $sgpr32
    S_ENDPGM 0

File: llvm/test/CodeGen/AMDGPU/sgpr-spills-empty-block.mir (hunk 3 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/sgpr-spills-empty-block.mir:8]: Oh right, I never pushed the block reduction for MIR

**Comment by bcahoon** [comment]
Thanks, CD! Much appreciated. I've replaced the test case.

**Review by cdevadas — APPROVED** [review]
LGTM.

**Review by cdevadas — APPROVED** [review]

**Pull request merged**